torch/autograd/saved_variable_default_hooks.py

```diff
@@ -8,18 +8,49 @@ def reset_saved_tensors_default_hooks():
     torch._C._autograd._reset_default_hooks()
 
 class save_on_cpu(object):
+    """"Context-manager under which tensors saved by the forward pass will be
+    stored on cpu, then retrieved for backward.
+
+    When performing operations within this context manager, intermediary
+    results saved in the graph during the forward pass will be moved to CPU,
+    then copied back to the original device when needed for the backward pass.
+    If the graph was already on CPU, no tensor copy is performed.
+
+    Use this context-manager to tradeoff speed for less GPU memory usage.
+
+    Args:
+        pin_memory (bool): If ``True`` tensors will be saved to CPU pinned memory
+                           during packing and copied to GPU asynchronously during unpacking.
+                           Defaults to ``False``.
+                           Also see :ref:`cuda-memory-pinning`.
+
+
+    Example::
+
+        >>> a = torch.randn(5, requires_grad=True, device="cuda")
+        >>> b = torch.randn(5, requires_grad=True, device="cuda")
+        >>> c = torch.randn(5, requires_grad=True, device="cuda")
+        >>> d = a * b # a and b are saved in the graph (on GPU)
+        >>> with torch.autograd.graph.save_on_cpu():
+        ...    e = d * c # d and c are saved on CPU
+        >>> f = a * e # a and e are saved on GPU
+        >>> del a, b, c, d, e
+        >>> # the content of a, b, e are still alive on GPU
+        >>> # the content of c and d only live on CPU
```

Comment:
Might be good to have a .backward() somewhere to demonstrate that the backward pass still works and that the autograd computation graph is being moved to cpu

Also I found this example a bit difficult to read but I'm not sure what a better example would be

Reply:
Hopefully, this is more readable:

```python
 a = torch.randn(5, requires_grad=True, device="cuda")
 b = torch.randn(5, requires_grad=True, device="cuda")
 c = torch.randn(5, requires_grad=True, device="cuda")

def f(a, b, c):
    prod_1 = a * b           # a and b are saved on GPU
    with torch.autograd.graph.save_on_cpu():
        prod_2 = prod_1 * c  # prod_1 and c are saved on CPU
    y = prod_2 * a           # prod_2 and a are saved on GPU
    return y
    
y = f(a, b, c)
del a, b, c  # for illustration only 
# the content of a, b, and prod_2 are still alive on GPU
# the content of prod_1 and c only live on CPU
y.sum().backward()  # all CPU tensors are moved back to GPU
```

Reply:
Yes, that looks better to me, thanks for taking the effort to make it easier to read!